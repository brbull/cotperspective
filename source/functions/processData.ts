import csv from 'csv-parse/lib/sync'
import pick from 'lodash/pick'
import { propertiesToSelect } from '../data/propertiesToSelect'
import { COTData, CSVData } from '../data/types'

/**
 * Transforms csv string into js object, grouped first by exchange, then by market name.
 * This grouping makes view rendering easier (mapping through the values)
 */
export const processData = (csvContent: string): COTData =>
  csv(csvContent, { trim: true, columns: true, skip_empty_lines: true }).reduce(
    (data: COTData, csvData: CSVData) => {
      const marketNameSplit = csvData['Market and Exchange Names'].split(' - ')
      const market = marketNameSplit[0].trim()
      const exchange = marketNameSplit[1].trim()
      // Set initial values
      if (!data[exchange]) data[exchange] = {}
      if (!data[exchange][market]) data[exchange][market] = []
      // Filter csv data, and push to result
      data[exchange][market].push({
        market,
        exchange,
        ...pick(csvData, propertiesToSelect)
      })
      return data
    },
    {}
  )
