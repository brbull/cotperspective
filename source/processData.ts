import csv from 'csv-parse/lib/sync'
import pick from 'lodash/pick'

const propertiesToSelect = <const> [
  'Market and Exchange Names',
  'As of Date in Form YYYY-MM-DD',
  'Open Interest (All)',
  'Noncommercial Positions-Long (All)',
  'Noncommercial Positions-Short (All)',
  'Commercial Positions-Long (All)',
  'Commercial Positions-Short (All)',
  'Change in Noncommercial-Long (All)',
  'Change in Noncommercial-Short (All)',
  'Change in Commercial-Long (All)',
  'Change in Commercial-Short (All)',
  '% of OI-Noncommercial-Long (All)',
  '% of OI-Noncommercial-Short (All)',
  '% of OI-Commercial-Long (All)',
  '% of OI-Commercial-Short (All)'
]

export type CSVData = Record<typeof propertiesToSelect[number], string>

export interface FormattedCSVData extends CSVData {
  market: string,
  exchange: string
}

export type COTData = Record<string, Record<string, FormattedCSVData[]>>

/**
 * Transforms csv string into js object, grouped first by exchange, then by market name.
 * This grouping makes view rendering easier (mapping through the values)
 */
export const processData = (csvContent: string): COTData => {
  return csv(csvContent, { trim: true, columns: true, skip_empty_lines: true })
    .reduce((data: COTData, csvData: CSVData) => {
      const marketNameSplit = csvData['Market and Exchange Names'].split(' - ')
      const market = marketNameSplit[0].trim()
      const exchange = marketNameSplit[1].trim()

      if (!data[exchange]) data[exchange] = {}
      if (!data[exchange][market]) data[exchange][market] = []

      data[exchange][market].push({ market, exchange, ...pick(csvData, propertiesToSelect) })

      return data
    }, {})
}
