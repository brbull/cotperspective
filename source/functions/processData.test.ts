import { processData } from './processData'
import { testCSV } from './testUtils/testData'

/**
 * Verify correct processing of test CSV data
 */
test('processData Snapshot', () => {
  const processedData = processData(testCSV)
  expect(processedData).toMatchSnapshot()
})
