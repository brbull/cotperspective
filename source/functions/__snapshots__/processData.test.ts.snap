// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processData Snapshot 1`] = `
Object {
  "CHICAGO BOARD OF TRADE": Object {
    "CORN": Array [
      Object {
        "% of OI-Commercial-Long (All)": "44.3",
        "% of OI-Commercial-Short (All)": "49.2",
        "% of OI-Noncommercial-Long (All)": "18.5",
        "% of OI-Noncommercial-Short (All)": "10.1",
        "As of Date in Form YYYY-MM-DD": "2020-09-15",
        "Change in Commercial-Long (All)": "26294",
        "Change in Commercial-Short (All)": "53726",
        "Change in Noncommercial-Long (All)": "21514",
        "Change in Noncommercial-Short (All)": "-15842",
        "Commercial Positions-Long (All)": "813715",
        "Commercial Positions-Short (All)": "904330",
        "Market and Exchange Names": "CORN - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "340202",
        "Noncommercial Positions-Short (All)": "186502",
        "Open Interest (All)": "1837646",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "CORN",
      },
      Object {
        "% of OI-Commercial-Long (All)": "45.0",
        "% of OI-Commercial-Short (All)": "48.6",
        "% of OI-Noncommercial-Long (All)": "18.2",
        "% of OI-Noncommercial-Short (All)": "11.6",
        "As of Date in Form YYYY-MM-DD": "2020-09-08",
        "Change in Commercial-Long (All)": "16197",
        "Change in Commercial-Short (All)": "32417",
        "Change in Noncommercial-Long (All)": "6273",
        "Change in Noncommercial-Short (All)": "-18605",
        "Commercial Positions-Long (All)": "787421",
        "Commercial Positions-Short (All)": "850604",
        "Market and Exchange Names": "CORN - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "318688",
        "Noncommercial Positions-Short (All)": "202343",
        "Open Interest (All)": "1751115",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "CORN",
      },
    ],
    "MINI SOYBEANS": Array [
      Object {
        "% of OI-Commercial-Long (All)": "0.0",
        "% of OI-Commercial-Short (All)": "27.0",
        "% of OI-Noncommercial-Long (All)": "85.5",
        "% of OI-Noncommercial-Short (All)": "9.5",
        "As of Date in Form YYYY-MM-DD": "2020-09-15",
        "Change in Commercial-Long (All)": "-1",
        "Change in Commercial-Short (All)": "1199",
        "Change in Noncommercial-Long (All)": "2221",
        "Change in Noncommercial-Short (All)": "571",
        "Commercial Positions-Long (All)": "8",
        "Commercial Positions-Short (All)": "5680",
        "Market and Exchange Names": "MINI SOYBEANS - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "17955",
        "Noncommercial Positions-Short (All)": "1987",
        "Open Interest (All)": "20999",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "MINI SOYBEANS",
      },
      Object {
        "% of OI-Commercial-Long (All)": "0.0",
        "% of OI-Commercial-Short (All)": "24.3",
        "% of OI-Noncommercial-Long (All)": "85.2",
        "% of OI-Noncommercial-Short (All)": "7.7",
        "As of Date in Form YYYY-MM-DD": "2020-09-08",
        "Change in Commercial-Long (All)": "-7",
        "Change in Commercial-Short (All)": "345",
        "Change in Noncommercial-Long (All)": "1966",
        "Change in Noncommercial-Short (All)": "558",
        "Commercial Positions-Long (All)": "9",
        "Commercial Positions-Short (All)": "4481",
        "Market and Exchange Names": "MINI SOYBEANS - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "15734",
        "Noncommercial Positions-Short (All)": "1416",
        "Open Interest (All)": "18465",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "MINI SOYBEANS",
      },
      Object {
        "% of OI-Commercial-Long (All)": "0.1",
        "% of OI-Commercial-Short (All)": "25.3",
        "% of OI-Noncommercial-Long (All)": "84.1",
        "% of OI-Noncommercial-Short (All)": "5.2",
        "As of Date in Form YYYY-MM-DD": "2020-09-01",
        "Change in Commercial-Long (All)": ".",
        "Change in Commercial-Short (All)": ".",
        "Change in Noncommercial-Long (All)": ".",
        "Change in Noncommercial-Short (All)": ".",
        "Commercial Positions-Long (All)": "16",
        "Commercial Positions-Short (All)": "4136",
        "Market and Exchange Names": "MINI SOYBEANS - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "13768",
        "Noncommercial Positions-Short (All)": "858",
        "Open Interest (All)": "16372",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "MINI SOYBEANS",
      },
    ],
    "SOYBEANS": Array [
      Object {
        "% of OI-Commercial-Long (All)": "42.7",
        "% of OI-Commercial-Short (All)": "60.6",
        "% of OI-Noncommercial-Long (All)": "23.7",
        "% of OI-Noncommercial-Short (All)": "4.2",
        "As of Date in Form YYYY-MM-DD": "2020-09-15",
        "Change in Commercial-Long (All)": "23608",
        "Change in Commercial-Short (All)": "37287",
        "Change in Noncommercial-Long (All)": "16576",
        "Change in Noncommercial-Short (All)": "-3406",
        "Commercial Positions-Long (All)": "489616",
        "Commercial Positions-Short (All)": "695451",
        "Market and Exchange Names": "SOYBEANS - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "271617",
        "Noncommercial Positions-Short (All)": "48618",
        "Open Interest (All)": "1147625",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "SOYBEANS",
      },
      Object {
        "% of OI-Commercial-Long (All)": "43.6",
        "% of OI-Commercial-Short (All)": "61.6",
        "% of OI-Noncommercial-Long (All)": "23.9",
        "% of OI-Noncommercial-Short (All)": "4.9",
        "As of Date in Form YYYY-MM-DD": "2020-09-08",
        "Change in Commercial-Long (All)": "18330",
        "Change in Commercial-Short (All)": "18954",
        "Change in Noncommercial-Long (All)": "10082",
        "Change in Noncommercial-Short (All)": "3157",
        "Commercial Positions-Long (All)": "466008",
        "Commercial Positions-Short (All)": "658163",
        "Market and Exchange Names": "SOYBEANS - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "255041",
        "Noncommercial Positions-Short (All)": "52024",
        "Open Interest (All)": "1069089",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "SOYBEANS",
      },
    ],
    "WHEAT-SRW": Array [
      Object {
        "% of OI-Commercial-Long (All)": "35.7",
        "% of OI-Commercial-Short (All)": "37.4",
        "% of OI-Noncommercial-Long (All)": "25.9",
        "% of OI-Noncommercial-Short (All)": "20.1",
        "As of Date in Form YYYY-MM-DD": "2020-09-15",
        "Change in Commercial-Long (All)": "3184",
        "Change in Commercial-Short (All)": "-8272",
        "Change in Noncommercial-Long (All)": "-7382",
        "Change in Noncommercial-Short (All)": "3969",
        "Commercial Positions-Long (All)": "165393",
        "Commercial Positions-Short (All)": "173120",
        "Market and Exchange Names": "WHEAT-SRW - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "119823",
        "Noncommercial Positions-Short (All)": "93044",
        "Open Interest (All)": "462819",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "WHEAT-SRW",
      },
      Object {
        "% of OI-Commercial-Long (All)": "34.8",
        "% of OI-Commercial-Short (All)": "38.9",
        "% of OI-Noncommercial-Long (All)": "27.3",
        "% of OI-Noncommercial-Short (All)": "19.1",
        "As of Date in Form YYYY-MM-DD": "2020-09-08",
        "Change in Commercial-Long (All)": "5831",
        "Change in Commercial-Short (All)": "603",
        "Change in Noncommercial-Long (All)": "-6720",
        "Change in Noncommercial-Short (All)": "2201",
        "Commercial Positions-Long (All)": "162209",
        "Commercial Positions-Short (All)": "181392",
        "Market and Exchange Names": "WHEAT-SRW - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "127204",
        "Noncommercial Positions-Short (All)": "89075",
        "Open Interest (All)": "466640",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "WHEAT-SRW",
      },
      Object {
        "% of OI-Commercial-Long (All)": "34.0",
        "% of OI-Commercial-Short (All)": "39.3",
        "% of OI-Noncommercial-Long (All)": "29.1",
        "% of OI-Noncommercial-Short (All)": "18.9",
        "As of Date in Form YYYY-MM-DD": "2020-09-01",
        "Change in Commercial-Long (All)": "-2992",
        "Change in Commercial-Short (All)": "25405",
        "Change in Noncommercial-Long (All)": "19159",
        "Change in Noncommercial-Short (All)": "-12653",
        "Commercial Positions-Long (All)": "156378",
        "Commercial Positions-Short (All)": "180790",
        "Market and Exchange Names": "WHEAT-SRW - CHICAGO BOARD OF TRADE",
        "Noncommercial Positions-Long (All)": "133924",
        "Noncommercial Positions-Short (All)": "86874",
        "Open Interest (All)": "460359",
        "exchange": "CHICAGO BOARD OF TRADE",
        "market": "WHEAT-SRW",
      },
    ],
  },
  "ICE FUTURES ENERGY DIV": Object {
    "RGGI VINTAGE 2020": Array [
      Object {
        "% of OI-Commercial-Long (All)": "87.5",
        "% of OI-Commercial-Short (All)": "89.3",
        "% of OI-Noncommercial-Long (All)": "10.3",
        "% of OI-Noncommercial-Short (All)": "9.2",
        "As of Date in Form YYYY-MM-DD": "2020-09-15",
        "Change in Commercial-Long (All)": "-28",
        "Change in Commercial-Short (All)": "119",
        "Change in Noncommercial-Long (All)": "147",
        "Change in Noncommercial-Short (All)": "-2",
        "Commercial Positions-Long (All)": "23748",
        "Commercial Positions-Short (All)": "24236",
        "Market and Exchange Names": "RGGI VINTAGE 2020 - ICE FUTURES ENERGY DIV",
        "Noncommercial Positions-Long (All)": "2797",
        "Noncommercial Positions-Short (All)": "2492",
        "Open Interest (All)": "27140",
        "exchange": "ICE FUTURES ENERGY DIV",
        "market": "RGGI VINTAGE 2020",
      },
      Object {
        "% of OI-Commercial-Long (All)": "88.1",
        "% of OI-Commercial-Short (All)": "89.4",
        "% of OI-Noncommercial-Long (All)": "9.8",
        "% of OI-Noncommercial-Short (All)": "9.2",
        "As of Date in Form YYYY-MM-DD": "2020-09-08",
        "Change in Commercial-Long (All)": "-368",
        "Change in Commercial-Short (All)": "-373",
        "Change in Noncommercial-Long (All)": "120",
        "Change in Noncommercial-Short (All)": "-102",
        "Commercial Positions-Long (All)": "23776",
        "Commercial Positions-Short (All)": "24117",
        "Market and Exchange Names": "RGGI VINTAGE 2020 - ICE FUTURES ENERGY DIV",
        "Noncommercial Positions-Long (All)": "2650",
        "Noncommercial Positions-Short (All)": "2494",
        "Open Interest (All)": "26978",
        "exchange": "ICE FUTURES ENERGY DIV",
        "market": "RGGI VINTAGE 2020",
      },
    ],
  },
  "MINNEAPOLIS GRAIN EXCHANGE": Object {
    "WHEAT-HRSpring": Array [
      Object {
        "% of OI-Commercial-Long (All)": "51.4",
        "% of OI-Commercial-Short (All)": "47.5",
        "% of OI-Noncommercial-Long (All)": "24.0",
        "% of OI-Noncommercial-Short (All)": "27.4",
        "As of Date in Form YYYY-MM-DD": "2020-09-15",
        "Change in Commercial-Long (All)": "-295",
        "Change in Commercial-Short (All)": "971",
        "Change in Noncommercial-Long (All)": "1175",
        "Change in Noncommercial-Short (All)": "-342",
        "Commercial Positions-Long (All)": "29821",
        "Commercial Positions-Short (All)": "27541",
        "Market and Exchange Names": "WHEAT-HRSpring - MINNEAPOLIS GRAIN EXCHANGE",
        "Noncommercial Positions-Long (All)": "13939",
        "Noncommercial Positions-Short (All)": "15891",
        "Open Interest (All)": "57961",
        "exchange": "MINNEAPOLIS GRAIN EXCHANGE",
        "market": "WHEAT-HRSpring",
      },
      Object {
        "% of OI-Commercial-Long (All)": "53.6",
        "% of OI-Commercial-Short (All)": "47.3",
        "% of OI-Noncommercial-Long (All)": "22.7",
        "% of OI-Noncommercial-Short (All)": "28.9",
        "As of Date in Form YYYY-MM-DD": "2020-09-08",
        "Change in Commercial-Long (All)": "-1850",
        "Change in Commercial-Short (All)": "2166",
        "Change in Noncommercial-Long (All)": "-373",
        "Change in Noncommercial-Short (All)": "-4648",
        "Commercial Positions-Long (All)": "30116",
        "Commercial Positions-Short (All)": "26570",
        "Market and Exchange Names": "WHEAT-HRSpring - MINNEAPOLIS GRAIN EXCHANGE",
        "Noncommercial Positions-Long (All)": "12764",
        "Noncommercial Positions-Short (All)": "16233",
        "Open Interest (All)": "56152",
        "exchange": "MINNEAPOLIS GRAIN EXCHANGE",
        "market": "WHEAT-HRSpring",
      },
    ],
  },
}
`;
